include/lbann/callbacks/callback.hpp

```diff
@@ -83,10 +84,14 @@ class callback_base {
   /** @name Modifiers */
   ///@{
 
-  /** @brief Called once to set up the callback (after all layers are
-   *         set up).
+  /** @brief Called once to set up the callback on the trainer
+   */
+  virtual void setup(trainer *t) {};
+
+  /** @brief Called once to set up the callback on the model
+   *         (after all layers are set up).
    */
-  virtual void setup(model *m) {};
+  virtual void setup(model *m, const std::string& trainer_name) {};
```

Comment:
This is more a general comment than a review: This is highlighting to me the poor design of callbacks. In particular, the API suggests that you can pass in any ol' `model*` and whatever the callback does will "just work". But many callbacks keep state and some even keep model-specific state. If the intention is that callbacks are "logically stateless", we have several callbacks that defy logic but the API is fine. If the intention is that callbacks can have state, there shouldn't be a `model*` to any API call except `setup()`, which would cache its input `model*` and use that when its hook functions are called. This could be managed in the base class with no significant changes to the derived classes.

The reason this change draws this API inconsistency to mind is that there should be no need for the `trainer_name` argument here as models are not shared among trainers. That is, this is a well-defined function: `trainer& get_trainer(model const&)` and we shouldn't need to expand the callback API with redundant information.

Reply:
Okay, I have fixed the issue with passing the trainer name around.  I am not sure what to do about the model because in this new world order a trainer has a single checkpoint callback, but it could be applied to multiple models.  Thus it is function with respect the model and stateful with respect to the trainer.

Reply:
> Thus it is function with respect the model and stateful with respect to the trainer.

This is the case in which it makes sense to have a `model*` parameter (technically it should be a `model&` since null models are not allowed). So I'm ok with the `model*` in this case. It's the "model callbacks" that should not have it... But that's a longer discussion.